feat: Query and display basic web vitals data

rafaeelaudibert opened the pull request:
We're now grabbing the P90 data and displaying as a chart. We've still got to:
    - Display them individually rather than grouped
    - Display information about what each metric means
    - Allow customers to choose between P75, P90, P95, P99
    - Display a range to explain what's the expected value for bad/good/excellent
    - Display an aggregate metric
    - Display a valid error when our query fails for lack of numbers (`quantile` returns a bad error if it's empty)

![image](https://github.com/user-attachments/assets/4fddae72-7e1f-4972-af51-35798d8098cb)

rafaeelaudibert committed:
refactor: Simplify `Query` component

By extracting all of the `query` prop handling to a separate component and keeping only the render inside a `DoQuery` component we can also make use of `useMemo` to slightly optimize our component avoiding several rerenders

rafaeelaudibert committed:
docs: Fix typo

rafaeelaudibert committed:
refactor: Remove pie chart handling dead code

We're not displaying any Pie chart in the Web analytics view, so let's remove this dead code

rafaeelaudibert committed:
refactor: Remove LCP from base Web Analytics view

Since we'll be displaying LCP under the new Web Analytics page, we're ok removing it from the main Web Analytics page. This should also make the queries slightly faster because we don't need to compute a heavy quantile

rafaeelaudibert committed:
feat: Query and display basic web vitals data from server

We're now grabbing the P90 data and displaying as a chart. We've still got to:
- Display them individually rather than grouped
- Display information about what each metric means
- Display a range to explain what's the expected value for bad/good/excellent
- Display an aggregate metric
- Display a valid error when our query fails for lack of numbers (`quantile` returns a bad error if it's empty)

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -711,23 +714,72 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
-                            tileId: TileId.OVERVIEW,
+                            tileId: TileId.WEB_VITALS,
+
+                            // TODO: Do we need title and docs? Likely not
```
Will be removing this once we have UI that's closer to the end goal

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -711,23 +714,72 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
+                                            event: '$web_vitals',
+                                            name: '$web_vitals',
+                                            custom_name: 'INP P90',
+                                            math: PropertyMathType.P90,
```
@robbie-c UI crashes real bad if we don't have Web Vitals data because `percentile(n)(value)` for an empty list of values is an error 🤷🏻. What's the best way to handle this kind of error?

I'm reusing `TrendsQuery`, my solution would be to create a child query in the backend that detects that and returns a more reasonable error. Thoughts?

robbie-c replied:
Weird, I just ran this query in metabase
`select quantile(0.5)(timestamp) from events where team_id = 2 and false`
and it returned a value (the zero timestamp).

Similarly 
`select quantile(0.5)(toInt64OrNull(events.$session_id)) from events where team_id = 2 and false`
gave me null.

Can you share the clickhouse code that's being generated?

rafaeelaudibert replied:
Maybe it's not that's empty, but rather a property that doesn't exist?

This is an example of me querying a non-existent `JJJ` web vital (which is basically a thing for INP when it hasn't been received yet). This is the error I get back: `Illegal type String of argument for aggregate function quantile.`

```sql
SELECT
    arrayMap(number -> plus(toStartOfDay(assumeNotNull(parseDateTime64BestEffortOrNull('2024-12-10 00:00:00', 6, 'UTC'))), toIntervalDay(number)), range(0, plus(coalesce(dateDiff('day', toStartOfDay(assumeNotNull(parseDateTime64BestEffortOrNull('2024-12-10 00:00:00', 6, 'UTC'))), toStartOfDay(assumeNotNull(parseDateTime64BestEffortOrNull('2025-01-09 23:59:59', 6, 'UTC'))))), 1))) AS date,
    arrayMap(_match_date -> arraySum(arraySlice(groupArray(ifNull(count, 0)), indexOf(groupArray(day_start) AS _days_for_count, _match_date) AS _index, plus(minus(arrayLastIndex(x -> ifNull(equals(x, _match_date), isNull(x) and isNull(_match_date)), _days_for_count), _index), 1))), date) AS total
FROM
    (SELECT
        sum(total) AS count,
        day_start AS day_start
    FROM
        (SELECT
            quantile(0.9)(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(e.properties, '$web_vitals_JJJ_value'), ''), 'null'), '^"|"$', '')) AS total,
            toStartOfDay(toTimeZone(e.timestamp, 'UTC')) AS day_start
        FROM
            events AS e SAMPLE 1
        WHERE
            and(equals(e.team_id, 1), greaterOrEquals(toTimeZone(e.timestamp, 'UTC'), toStartOfDay(assumeNotNull(parseDateTime64BestEffortOrNull('2024-12-10 00:00:00', 6, 'UTC')))), lessOrEquals(toTimeZone(e.timestamp, 'UTC'), assumeNotNull(parseDateTime64BestEffortOrNull('2025-01-09 23:59:59', 6, 'UTC'))), equals(e.event, '$web_vitals'), or(isNull(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(e.properties, '$browser'), ''), 'null'), '^"|"$', '')), not(JSONHas(e.properties, '$browser'))))
        GROUP BY
            day_start)
    GROUP BY
        day_start
    ORDER BY
        day_start ASC)
ORDER BY
    arraySum(total) DESC
LIMIT 50000 SETTINGS readonly=2, max_execution_time=600, allow_experimental_object_type=1, format_csv_allow_double_quotes=0, max_ast_elements=4000000, max_expanded_ast_elements=4000000, max_bytes_before_external_group_by=0
```

robbie-c replied:
Oh right, so it's more a problem that HogQL is giving the wrong type for this property if it doesn't exist, so is not converting it to a number

robbie-c replied:
This might be a bigger project that you've just wandered into, but IMO all built-in properties should use hard-code types. This would solve your problem and potentially fix some other weird customer support issues

rafaeelaudibert replied:
Oh, I understand the above now. I'm querying all of them at the same time but we actually generate 4 different Clickhouse queries. Only that JJJ one fails, the others succeed (as expected)

robbie-c replied:
I'm afraid I don't know. I would guess that this happens somewhere in HogQL, and that it would involve pulling in PropertyDefinitions from the django model of that name, but to be sure I'd need to run through it in a debugger or spend more time staring at the code.

rafaeelaudibert replied:
I'll choose the lazy but faster route and solve this at a later stage in the Web Vitals project, in the meantime I'll see if someone in #dev knows which weeds I need to get myself into, thanks!

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

rafaeelaudibert committed:
revert: Go back to single component

`useMemo` was slightly weird with the previous approach, let's go back to a single component

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

robbie-c commented on frontend/src/queries/Query/Query.tsx:
```diff
@@ -71,16 +72,92 @@ export function Query<Q extends Node>(props: QueryProps<Q>): JSX.Element | null
     const setQuery = propsSetQuery ?? localSetQuery
 
-    const queryContext = props.context || {}
+    const queryContext = useMemo(() => props.context || {}, [props.context])
```
Do you know if this will gain you anything? It's just an object key lookup; I would guess that `useMemo` is doing more than that behind the scenes.

rafaeelaudibert replied:
This is because I wanna use `queryContext` as a dependency for the `useMemo` below. If we keep what we had before, then `{}` would always be a new object, therefore we wouldn't be caching anything

robbie-c replied:
Ah true, yes it would help a lot in that case!

robbie-c commented on frontend/src/queries/Query/Query.tsx:
```diff
@@ -71,16 +72,92 @@ export function Query<Q extends Node>(props: QueryProps<Q>): JSX.Element | null
     const uniqueKey =
         props.uniqueKey ?? (props.context?.insightProps && insightVizDataNodeKey(props.context.insightProps))
 
+    const component = useMemo(() => {
```
I have the same question here. We're doing more work than the key lookup above, but a few if statements and creating a react element are not expensive computations. Does it prevent some unnecessary re-renders?

rafaeelaudibert replied:
Done, removed this

robbie-c commented on frontend/src/queries/nodes/WebOverview/WebOverview.tsx:
```diff
@@ -59,11 +59,10 @@ export function WebOverview(props: {
             {samplingRate && !(samplingRate.numerator === 1 && (samplingRate.denominator ?? 1) === 1) ? (
                 <LemonBanner type="info" className="my-4">
-                    These results using a sampling factor of {samplingRate.numerator}
+                    These results are using a sampling factor of {samplingRate.numerator}
```
oops

Though we don't actually use sampling here as it's not numerically sound

robbie-c replied:
It's just never turned on, I built the feature but the problems with soundness were too much for it to be useful.

We'd need to sample by distinct id for us to be able to use it

rafaeelaudibert replied:
Aah, gotcha. Does it make sense to remove it then? I always have to wrangle with the queries to make them work with the eventual `_.unsample` call

robbie-c replied:
Potentially yeah - I know there have been discussions around changing how this works so that we *can* sample by person ID, but I don't think there's anything concrete and I have my doubts as to whether we'd do it soon.

For extra context:
Clickhouse doesn't really have a normal index in the way that e.g. a Postgres user would consider them. You only have the order that things are laid out on disk, so when I say "index" to talk about clickhouse, that's what I mean.
To make sampling fast, the sample key (or `hash(sample_key)` must be part of the index. This allows clickhouse to skip reading huge chunks of data, and without it, sampling doesn't gain you much performance.
Updating values in clickhouse is slow but doable if the the thing you update is not part of the index. Changing a value that is part of the index means that rows need to be laid out in a different order on disk, which is very expensive.
Our person model means that the person ID of an event needs to be updated, e.g. when an $identify event causes 2 users to be merged.

robbie-c commented on frontend/src/queries/schema.ts:
```diff
@@ -1856,7 +1856,6 @@ interface WebAnalyticsQueryBase<R extends Record<string, any>> extends DataNode<
 
 export interface WebOverviewQuery extends WebAnalyticsQueryBase<WebOverviewQueryResponse> {
     kind: NodeKind.WebOverviewQuery
-    includeLCPScore?: boolean
```
Weak opinion: but you could keep this, and have it link to web vitals. If we are going to remove it I'd wait until core web vitals is live, and disable it with the web vitals FF

rafaeelaudibert replied:
My problem with this value here is that it is not the same value as people will see in the Web Vitals UI because you're grabbing this from the session and not from the `$web_vitals` events like I'm doing. I'm afraid keeping this will make it more confusing than if we simply removed it.

As for keeping it and hiding it when people enable the flag: Yeah, I could do that if you felt strongly about it, but I also believe that we should not be displaying data that's not 100% true and/or doesn't promote any kind of insight so I'll keep this change for now

rafaeelaudibert committed:
refactor: Revert refactor to Query

We don't have a case for improved performance so let's revert this now

rafaeelaudibert committed:
Merge branch 'master' into display-web-vitals-metrics-on-ui
